lms/djangoapps/grades/tests/test_signals.py

```diff
@@ -148,3 +158,74 @@ def test_score_reset_bad_user(self):
         self.get_user_mock = self.setup_patch('lms.djangoapps.grades.signals.user_by_anonymous_id', None)
         submissions_score_reset_handler(None, **SUBMISSION_RESET_KWARGS)
         self.signal_mock.assert_not_called()
+
+
+@ddt.ddt
+class ScoreChangedUpdatesSubsectionGradeTest(ModuleStoreTestCase):
+    """
+    Ensures that upon SCORE_CHANGED signals, the handler
+    initiates an update to the affected subsection grade.
+    """
+    def setUp(self):
+        super(ScoreChangedUpdatesSubsectionGradeTest, self).setUp()
+
+    def setUpCourseAndUser(self, enable_subsection_grades):
+        """
+        Configures the course for this test.
+        """
+        self.user = UserFactory()
+        self.course = CourseFactory.create(
+            org='edx',
+            name='course',
+            run='run',
+            metadata={'enable_subsection_grades_saved': enable_subsection_grades})
+        
+        self.chapter = ItemFactory.create(parent=self.course, category="chapter", display_name="Chapter")
+        self.sequential = ItemFactory.create(parent=self.chapter, category='sequential', display_name="Open Sequential")
+        self.problem = ItemFactory.create(parent=self.sequential, category='problem', display_name='problem')
+
+        self.score_changed_kwargs = {
+            'points_possible': 10,
+            'points_earned': 5,
+            'user_id': self.user.id,
+            'course_id': unicode(self.course.id),
+            'usage_id': unicode(self.problem.location),
+        }
+
+    @ddt.data(ModuleStoreEnum.Type.mongo, ModuleStoreEnum.Type.split)
+    def test_subsection_grade_updated_on_signal(self, default_store):
+        with self.store.default_store(default_store):
+            self.setUpCourseAndUser(True)
+            with check_mongo_calls(2) and self.assertNumQueries(18):
```

Comment:
Here's my quick take on what's happening:
```
1. get user information
2. get user access role
3. get student module
4. get anonymous user id
    1. doesn’t exist, so savepoint-insert-release (+3) here
5. get saved submissions scores
6. get persistentsubsectiongrade
7. savepoint
    1. get visibleblocks
        1. doesn’t exist, savepoint-insert-release (+3)
    2. ***get visible blocks again
    3. ***get persistentsubsectiongrade again
    4. insert into PSG
8. release
```

creating an anonymous user id and creating a new visible blocks instance aren't actions we expect to see on every request, so that would save us 6 queries.

The other low-hanging fruit are the two lines I flagged with `***` - they seem to be re-requesting data we should already have.